pkg/gatewayserver/io/io.go

```diff
@@ -103,27 +106,22 @@ type Connection struct {
 	fps              *frequencyplans.Store
 	scheduler        *scheduling.Scheduler
 	rtts             *rtts
+	addr             *ttnpb.GatewayRemoteAddress
 
 	upCh     chan *ttnpb.GatewayUplinkMessage
 	downCh   chan *ttnpb.DownlinkMessage
 	statusCh chan *ttnpb.GatewayStatus
 	txAckCh  chan *ttnpb.TxAcknowledgment
 
-	statsChangedCh chan struct{}
-	locCh          chan struct{}
-
-	versionInfoCh chan struct{}
-
-	lastUplink            *uplinkMessage
-	lastRepeatUpEventTime time.Time
-
-	addr *ttnpb.GatewayRemoteAddress
+	statsChangedCh       chan struct{}
+	locChangedCh         chan struct{}
+	versionInfoChangedCh chan struct{}
 }
 
 type uplinkMessage struct {
-	payload   []byte
-	frequency uint64
-	antennas  []uint32
+	payloadHash uint64
```

Comment:
Should we also store the data rate in order to avoid false positives on actual end device retransmissions ?

Reply:
Yes good one.

Reply:
There is no time window check for this logic here - the gateway may send the 'duplicate' even after 2 days as long.

The original PR is https://github.com/TheThingsNetwork/lorawan-stack/pull/4181 and the original issue is https://github.com/TheThingsNetwork/lorawan-stack/issues/4095 . I think that the time window logic never got implemented, and for the past years we may have been dropping some legitimate retransmissions which had different data rates. If we would compare data rates too, the only possible false positive is SF12 on the same frequency, which is less likely than the current frequency re-use chance of collision.

Reply:
I think I was assuming the duplicate check would be time-bound, as I wrote in #4095. Without diving in the code, we're not keeping track of all uplink messages in the connection for duplicate checks, right?

Reply:
> Without diving in the code, we're not keeping track of all uplink messages in the connection for duplicate checks, right?

Correct - we always compare the last processed uplink only with the pending uplink (pending approval, since we would reject an uplink which is identical).

Reply:
Oh okay. Yeah then we'd need to factor in data rate too. In fact, the 10 seconds I was talking about in #4095 wouldn't be enough for retransmissions either. I think I was assuming that a retransmission would only use another frequency, not a different data rate.